src/main/java/org/jenkinsci/plugins/durabletask/WindowsBatchScript.java

```diff
@@ -59,34 +86,100 @@ public String getScript() {
         if (launcher.isUnix()) {
             throw new IOException("Batch scripts can only be run on Windows nodes");
         }
+
+        FilePath nodeRoot = getNodeRoot(ws);
+        final Jenkins jenkins = Jenkins.get();
+        PluginWrapper durablePlugin = jenkins.getPluginManager().getPlugin("durable-task");
+        if (durablePlugin == null) {
+            throw new IOException("Unable to find durable task plugin");
+        }
+        String pluginVersion = StringUtils.substringBefore(durablePlugin.getVersion(), "-");
+        AgentInfo agentInfo = nodeRoot.act(new AgentInfo.GetAgentInfo(pluginVersion));
         BatchController c = new BatchController(ws);
 
-        String cmd;
+        List<String> launcherCmd = null;
+        if (FORCE_BINARY_WRAPPER && agentInfo.isBinaryCompatible()) {
```

Comment:
Could some of this logic be extracted into `FileMonitoringTask` and reused across the different subclasses? Or are there too many differences?

Reply:
Was about to move some of the logic out to `FileMonitoringTask`. A lot of code was shared, the only issue was that each shell type was using it in a slightly unique way 😛 
See c224d6ae001d22c984829428032465285f38bec1 for changes